Control plane mtu

fcrisciani opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Add the control plane MTU option in the daemon config. This will be useful in clusters with
MTU higher than 1500. The first user of this option will be libnetwork that will use it to
seed the networkDB library.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Add the control plane MTU option in the daemon config

**- A picture of a cute animal (not mandatory but encouraged)**

![funny-cute-animal-pics-part119-17-w630](https://user-images.githubusercontent.com/1823625/28192664-7e32cb5a-67ec-11e7-9eb0-2563d93ccbd9.jpg)

Commit:
Libnetwork vendoring

Signed-off-by: Flavio Crisciani <flavio.crisciani@docker.com>

fcrisciani commented:
Waiting for the vendoring of the libnetwork side, but will take CR or comments in the mean time

friism commented:
@fcrisciani will this help with this problem? https://github.com/moby/moby/issues/33596

fcrisciani commented:
@friism unfortunately nope, this is for control plane traffic, that is instead data plane traffic issue. Good to know though because gossip cap the packet at 1400, so this feature with Microsoft node most likely won't work till they fix it.

fcrisciani commented:
@friism looks like we support MTU on the overlay: https://github.com/docker/libnetwork/pull/1349, did they try this?

cpuguy83 commented on daemon/config/config.go:
```diff
@@ -168,6 +168,9 @@ type CommonConfig struct {
 	Experimental bool `json:"experimental"` // Experimental indicates whether experimental features should be exposed or not
+
+	// ControlPlaneMTU allows to specify the control plane MTU, this will allow to optimize the network use in some components
+	ControlPlaneMTU int `json:"control-plane-mtu,omitempty"`
```
`NetworkControlPlaneMTU`?

fcrisciani replied:
sure, I will update it with once the libnetwork side is ready

cpuguy83 commented on cmd/dockerd/config.go:
```diff
@@ -62,6 +62,8 @@ func installCommonConfigFlags(conf *config.Config, flags *pflag.FlagSet) {
 
 	flags.StringVar(&conf.MetricsAddress, "metrics-addr", "", "Set default address and port to serve the metrics api on")
 
+	flags.IntVar(&conf.ControlPlaneMTU, "control-plane-mtu", config.DefaultNetworkMtu, "Control plane MTU")
```
`network-control-plane-mtu`?

fcrisciani replied:
ok

Commit:
Allow to set the control plane MTU

Add daemon config to allow the user to specify the MTU of the control plane network.
The first user of this new parameter is actually libnetwork that can seed the
gossip with the proper MTU value allowing to pack multiple messages per UDP packet sent.
If the value is not specified or is lower than 1500 the logic will set it to the default.

Signed-off-by: Flavio Crisciani <flavio.crisciani@docker.com>

fcrisciani commented:
@cpuguy83 ready for a second pass

cpuguy83 reviewed (approved):
Code and Design LGTM

vieux commented:
LGTM

thaJeztah commented:
> If the value is not specified or is lower than 1500 the logic will set it to the default.

@fcrisciani should values lower than 1500 produce an error instead of silently ignoring? I can see it being unexpected if someone sets a lower value and it's not being used (or; if lower values become valid in future, they become active after updating to a newer version)

fcrisciani commented:
@thaJeztah in reality that is a comment that did not get updated, on the libnetwork side we decided to allow any configuration but we will log a warning (https://github.com/docker/libnetwork/blob/bump_17.07/config/config.go#L231)

```
if exp < 1500 {
     logrus.Warnf("Received a MTU of %d, this value is very low, the network control plane can misbehave", exp)
}
```
